pkg/bucket/bandwidth/measurement.go

```diff
@@ -0,0 +1,133 @@
+func newBucketMeasurement(initTime time.Time) *bucketMeasurement {
+	return &bucketMeasurement{
+		windowStart: initTime,
+	}
+}
+
+// incrementBytes add bytes reported for a bucket.
+func (m *bucketMeasurement) incrementBytes(bytes uint64) {
+	atomic.AddUint64(&m.cumulativeBytes, bytes)
+	atomic.AddUint64(&m.bytesSinceLastWindow, bytes)
+}
+
+// calculateWindow processes the measurements between fixed time slots.
+func (m *bucketMeasurement) calculateWindow(windowEnd time.Time) {
+	// Calculate aggregate avg bandwidth and exp window avg
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	defer func() {
+		m.windowStart = windowEnd
+		m.sumOfTimePerObjectInWindow = 0
+	}()
+
+	if windowEnd.Before(m.windowStart) {
+		return
+	}
+
+	duration := windowEnd.Sub(m.windowStart)
+
+	// Do not update cumulative time if no traffic
+	bytesSinceLastWindow := atomic.SwapUint64(&m.bytesSinceLastWindow, 0)
+	if bytesSinceLastWindow > 0 {
+		m.cumulativeTime += duration
+		m.sumOfTimePerObject += m.sumOfTimePerObjectInWindow
+		// If total time spend copying object is less than wall clock time, then use
+		// sum of object to avoid  under reporting bandwidth.
+		if m.sumOfTimePerObject > 0 && m.cumulativeTime > m.sumOfTimePerObject {
+			m.cumulativeTime = m.sumOfTimePerObject
+		}
+	}
+	if m.sumOfTimePerObjectInWindow > 0 && duration > m.sumOfTimePerObjectInWindow {
+		duration = m.sumOfTimePerObjectInWindow
+	}
+
+	if m.expMovingAvg == 0 && m.cumulativeTime == 0 {
+		// Should address initial calculation and should be fine for resuming from 0
+		m.expMovingAvg = float64(bytesSinceLastWindow) / duration.Seconds()
+		return
+	}
+
+	increment := float64(bytesSinceLastWindow) / duration.Seconds()
+	m.expMovingAvg = exponentialMovingAverage(betaBucket, m.expMovingAvg, increment)
+}
+
+// exponentialMovingAverage calculates the exponential moving average
+func exponentialMovingAverage(beta, previousAvg, incrementAvg float64) float64 {
+	return (1-beta)*incrementAvg + beta*previousAvg
+}
+
+// getAvgBytesPerSecond returns the bandwidth in bytes
+func (m *bucketMeasurement) getAvgBytesPerSecond() float64 {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	seconds := float64(m.cumulativeTime) / float64(time.Second)
+	if seconds == 0 {
+		return 0
+	}
+	return float64(atomic.LoadUint64(&m.cumulativeBytes)) / seconds
+}
+
+// getTotalBytes returns the total bytes observed for a bucket
+func (m *bucketMeasurement) getTotalBytes() uint64 {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	return atomic.LoadUint64(&m.cumulativeBytes)
+}
+
+// getExpMovingAvgBytesPerSecond returns the exponential moving average for the bucket in bytes
+func (m *bucketMeasurement) getExpMovingAvgBytesPerSecond() float64 {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	return m.expMovingAvg
+}
+
+func (m *bucketMeasurement) incrementPerObjectDuration(sub time.Duration) {
+	if sub < 0 {
+		return
+	}
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	m.sumOfTimePerObjectInWindow += sub
+}
```

Comment:
Duration is an int64 beneath.
```suggestion
func (m *bucketMeasurement) incrementPerObjectDuration(sub time.Duration) {
	if sub < 0 {
		return
	}
	atomic.AddInt64((*int64)(&m.sumOfTimePerObjectInWindow), int64(sub))
}
```

Reply:
👍🏻